cpp/src/arrow/util/future.h

```diff
@@ -691,34 +876,41 @@ struct Continue {
   }
 };
 
-template <typename T = detail::Empty>
+template <typename T = internal::Empty>
 util::optional<T> Break(T break_value = {}) {
   return util::optional<T>{std::move(break_value)};
 }
 
-template <typename T = detail::Empty>
+template <typename T = internal::Empty>
 using ControlFlow = util::optional<T>;
 
+template <typename T>
+void ForwardControlResult(const Result<ControlFlow<T>>& result, Future<T> sink) {
+  sink.MarkFinished(**result);
+}
+template <>
+inline void ForwardControlResult(const Result<ControlFlow<>>& result, Future<> sink) {
+  sink.MarkFinished();
```

Comment:
I'm a bit skeptical that we have to introduce this yet another result-to-future forwarding primitive.

Reply:
This has gone away.